Automattic/wp-calypso
# Full Site Editing: Add basic feature flags support

#### Changes proposed in this Pull Request

* Add a very basic query param based feature flags system working in both PHP and JS.

This implementation uses the `?fse_flags` query param that can contain a comma-separated list of flags, or nothing to clear the saved flags.
It supports both enabling (by default) and disabling flags (by prepending `-` to the flag name), like in Calypso. E.g.
```
?fse_flags=enabledFlag,-disabledFlag
```

Flags are saved into a cookie that expires when closing the browser, or when explicitly sending an empty `?fse_flags`.

The flags are then passed to JS via `wp_localize_script`.

To check them:

```php
// PHP
require_once 'path/to/lib/feature-flags/feature-flags.php';

$feature_flags = A8C_Full_Site_Editing_Feature_Flags::get_instance();

if ( $feature_flags->is_enabled( 'flag' ) ) { /* ... */ }
```

```es6
// JS
import isEnabled from 'path/to/lib/feature-flags';

if ( isEnabled( 'flag' ) ) { /* ... */ }
```

#### Notes

I've willingly avoided config files because I'm not sure how to add support for environment-based flags.
As I'm going on leave soon, feel free to commandeer or kill this PR as y'all see fit. 🙂 

#### Testing instructions



There are no feature flags implemented yet, but this is a nice way to test the PR nonetheless.

* Follow the [FSE readme](https://github.com/Automattic/wp-calypso/blob/e759e056360b72576e124989c4174afef78bc1db/apps/full-site-editing/README.md#local-development), and setup the plugin.
* Open the block editor and observe `fullSiteEditing.featureFlags` in the browser console (it should be an empty array).
* Add `?fse_flags=flag1,-flag2` to the URL and reload.
* Now `fullSiteEditing.featureFlags` should be `{ flag1: true, flag2: false }`.
* Change the query params and make sure the `fullSiteEditing.featureFlags` keeps updating as expected.
* Finally, add `?fse_flags` (empty).
* Make sure `fullSiteEditing.featureFlags` is an empty array again.

Fixes #33240
